src/core/service.c

```diff
@@ -2228,11 +2228,13 @@ static bool service_good(Service *s) {
         if (s->type == SERVICE_DBUS && !s->bus_name_good)
                 return false;
 
-        main_pid_ok = main_pid_good(s);
-        if (main_pid_ok > 0) /* It's alive */
-                return true;
-        if (main_pid_ok == 0) /* It's dead */
-                return false;
+        if (s->exit_type == SERVICE_EXIT_MAIN) {
+                main_pid_ok = main_pid_good(s);
```

Comment:
You can reduce the scope of `main_pid_ok`, i.e. move the declaration into the if block too.